KAFKA-14370: Simplify the ImageWriter API by adding freeze

cmccabe opened the pull request:
Simplify the ImageWriter API by adding ImageWriter#freeze() instead of requiring a call to an
overload of ImageWriter#close().

jsancio commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:
```diff
@@ -398,8 +398,11 @@ class BrokerMetadataListener(
         _image.write(writer, options)
-      } finally {
-        writer.close()
+        writer.close(true)
```
`MetadataImage::write` calls `ImageWriter::close(true)`. Can we remove that call? It if very difficult to read and understand code that closes `AutoCloseable` in different layers of the code.

cmccabe replied:
Yes, you are right that we're already calling `ImageWriter.close(true)` in `MetadataImage.write`

jsancio reviewed (commented):
Can you explain the issue that you are fixing here? For example, what test is failing without this change?

cmccabe committed:
KAFKA-14370: Simplify the ImageWriter API by adding freeze

Simplify the ImageWriter API by adding ImageWriter#freeze() instead of requiring a call to an
overload of ImageWriter#close().

cmccabe commented:
It's a bit confusing to have two variations of close(), so let's just have freeze() and then close().
